File: windows/Gutenkern.Core/KerningCompletion.cs
namespace Gutenkern;

public sealed record PairLine(string Key, string Display);

public sealed class RecipeSection
{
    public string Recipe { get; }
    public IReadOnlyList<string> Groups { get; }
    public IReadOnlyList<IReadOnlyList<PairLine>> PairGroups { get; }

    public RecipeSection(string recipe, IReadOnlyList<string> groups)
        : this(recipe, groups.Select(ParseBlock).ToList())
    {
    }

    public RecipeSection(string recipe, IReadOnlyList<IReadOnlyList<PairLine>> pairGroups)
    {
        Recipe = recipe;
        PairGroups = pairGroups;
        Groups = pairGroups
            .Select(group => string.Join("\n", group.Select(pair => pair.Display)))
            .ToList();
    }

    private static IReadOnlyList<PairLine> ParseBlock(string block) =>
        KerningCompletion.PairKeys(block).Select(line => new PairLine(line, line)).ToList();
}

public sealed class KerningCompletion
{
    public HashSet<string> Recipes { get; }
    public HashSet<string> Blocks { get; }

    public KerningCompletion(IEnumerable<string> recipes, IEnumerable<string> blocks)
    {
        Recipes = [.. recipes];
        Blocks = ExpandPairKeys(blocks);
    }

    public bool IsRecipeDone(string recipe, IReadOnlyList<RecipeSection> sections) =>
        AllGroupsDone(recipe, sections);

    public bool IsBlockDone(string block) => IsBlockDone(PairKeys(block));

    public bool IsBlockDone(IReadOnlyList<string> keys) =>
        keys.Count > 0 && keys.All(Blocks.Contains);

    public void ToggleRecipe(string recipe, IReadOnlyList<RecipeSection> sections)
    {
        var keys = PairKeys(recipe, sections);
        if (IsRecipeDone(recipe, sections))
        {
            Recipes.Remove(recipe);
            foreach (var key in keys)
            {
                Blocks.Remove(key);
            }
        }
        else
        {
            Recipes.Add(recipe);
            foreach (var key in keys)
            {
                Blocks.Add(key);
            }
        }

        SyncRecipes(sections);
    }

    public void ToggleBlock(string block, IReadOnlyList<RecipeSection> sections) =>
        ToggleKeys(PairKeys(block), sections);

    public void ToggleKeys(IReadOnlyList<string> keys, IReadOnlyList<RecipeSection> sections)
    {
        if (IsBlockDone(keys))
        {
            foreach (var key in keys)
            {
                Blocks.Remove(key);
            }
        }
        else
        {
            foreach (var key in keys)
            {
                Blocks.Add(key);
            }
        }

        SyncRecipes(sections);
    }

    public void ApplyDone(IEnumerable<string> done, IReadOnlyList<RecipeSection> sections)
    {
        Blocks.Clear();
        foreach (var key in ExpandPairKeys(done))
        {
            Blocks.Add(key);
        }

        SyncRecipes(sections);
    }

    public void Sync(IReadOnlyList<RecipeSection> sections)
    {
        var known = AllPairKeys(sections);
        Blocks.RemoveWhere(key => !known.Contains(key));
        foreach (var section in sections)
        {
            if (section.PairGroups.Count == 0)
            {
                continue;
            }

            var keys = PairKeys(section);
            var marked = keys.Count(Blocks.Contains);
            if (Recipes.Contains(section.Recipe) && marked == 0)
            {
                foreach (var key in keys)
                {
                    Blocks.Add(key);
                }
            }
        }

        SyncRecipes(sections);
    }

    public void SyncRecipes(IReadOnlyList<RecipeSection> sections)
    {
        foreach (var section in sections)
        {
            if (section.PairGroups.Count == 0)
            {
                continue;
            }

            if (PairKeys(section).All(Blocks.Contains))
            {
                Recipes.Add(section.Recipe);
            }
            else
            {
                Recipes.Remove(section.Recipe);
            }
        }
    }

    public static List<string> PairKeys(string block) =>
        block.Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries)
            .Select(CanonicalizePairKey)
            .Where(key => key.Length > 0)
            .ToList();

    public static List<string> PairKeys(RecipeSection section) =>
        section.PairGroups.SelectMany(group => group.Select(pair => pair.Key)).ToList();

    public static List<string> PairKeys(string recipe, IReadOnlyList<RecipeSection> sections)
    {
        foreach (var section in sections)
        {
            if (section.Recipe == recipe)
            {
                return PairKeys(section);
            }
        }

        return [];
    }

    public static HashSet<string> AllPairKeys(IReadOnlyList<RecipeSection> sections) =>
        sections.SelectMany(PairKeys).ToHashSet();

    public static HashSet<string> ExpandPairKeys(IEnumerable<string> blocks) =>
        blocks.SelectMany(PairKeys).ToHashSet();

    public static string CanonicalizePairKey(string key) =>
        key.Length == 0 || key.StartsWith('/') ? key : "/" + key;

    private bool AllGroupsDone(string recipe, IReadOnlyList<RecipeSection> sections)
    {
        var keys = PairKeys(recipe, sections);
        return keys.Count > 0 && keys.All(Blocks.Contains);
    }
}

File: macos/Gutenkern/Sources/GutenkernCore/KerningCompletion.swift
import Foundation

public struct PairLine: Equatable, Hashable, Sendable {
    public let key: String
    public let display: String

    public init(key: String, display: String) {
        self.key = key
        self.display = display
    }
}

public struct RecipeSection: Equatable, Sendable {
    public let recipe: String
    public let pairGroups: [[PairLine]]

    public var groups: [String] {
        pairGroups.map { $0.map(\.display).joined(separator: "\n") }
    }

    public init(recipe: String, groups: [String]) {
        self.recipe = recipe
        self.pairGroups = groups.map { block in
            KerningCompletion.pairKeys(in: block).map { PairLine(key: $0, display: $0) }
        }
    }

    public init(recipe: String, pairGroups: [[PairLine]]) {
        self.recipe = recipe
        self.pairGroups = pairGroups
    }
}

public struct KerningCompletion: Equatable, Sendable {
    public var recipes: Set<String>
    public var blocks: Set<String>

    public init(recipes: Set<String> = [], blocks: Set<String> = []) {
        self.recipes = recipes
        self.blocks = Self.expandPairKeys(blocks)
    }

    public func isRecipeDone(_ recipe: String, sections: [RecipeSection]) -> Bool {
        allGroupsDone(recipe, sections: sections)
    }

    public func isBlockDone(_ block: String) -> Bool {
        isBlockDone(keys: Self.pairKeys(in: block))
    }

    public func isBlockDone(keys: [String]) -> Bool {
        !keys.isEmpty && keys.allSatisfy { blocks.contains($0) }
    }

    public mutating func toggleRecipe(_ recipe: String, sections: [RecipeSection]) {
        let keys = Self.pairKeys(for: recipe, in: sections)
        if isRecipeDone(recipe, sections: sections) {
            recipes.remove(recipe)
            for key in keys {
                blocks.remove(key)
            }
        } else {
            recipes.insert(recipe)
            for key in keys {
                blocks.insert(key)
            }
        }
        syncRecipes(sections: sections)
    }

    public mutating func toggleBlock(_ block: String, sections: [RecipeSection]) {
        toggleKeys(Self.pairKeys(in: block), sections: sections)
    }

    public mutating func toggleKeys(_ keys: [String], sections: [RecipeSection]) {
        if isBlockDone(keys: keys) {
            for key in keys {
                blocks.remove(key)
            }
        } else {
            for key in keys {
                blocks.insert(key)
            }
        }
        syncRecipes(sections: sections)
    }

    public mutating func applyDone(_ done: Set<String>, sections: [RecipeSection]) {
        blocks = Self.expandPairKeys(done)
        syncRecipes(sections: sections)
    }

    public mutating func sync(sections: [RecipeSection]) {
        let known = Self.allPairKeys(sections)
        blocks = blocks.intersection(known)
        for section in sections where !section.pairGroups.isEmpty {
            let keys = Self.pairKeys(in: section)
            let marked = keys.reduce(into: 0) { count, key in
                if blocks.contains(key) {
                    count += 1
                }
            }
            if recipes.contains(section.recipe), marked == 0 {
                for key in keys {
                    blocks.insert(key)
                }
            }
        }
        syncRecipes(sections: sections)
    }

    public mutating func syncRecipes(sections: [RecipeSection]) {
        for section in sections where !section.pairGroups.isEmpty {
            if Self.pairKeys(in: section).allSatisfy({ blocks.contains($0) }) {
                recipes.insert(section.recipe)
            } else {
                recipes.remove(section.recipe)
            }
        }
    }

    public static func pairKeys(in block: String) -> [String] {
        block.split(whereSeparator: \.isNewline).compactMap { line in
            let key = canonicalizePairKey(String(line))
            return key.isEmpty ? nil : key
        }
    }

    public static func pairKeys(in section: RecipeSection) -> [String] {
        section.pairGroups.flatMap { $0.map(\.key) }
    }

    public static func pairKeys(for recipe: String, in sections: [RecipeSection]) -> [String] {
        sections.first { $0.recipe == recipe }.map { pairKeys(in: $0) } ?? []
    }

    public static func allPairKeys(_ sections: [RecipeSection]) -> Set<String> {
        Set(sections.flatMap { pairKeys(in: $0) })
    }

    public static func expandPairKeys(_ blocks: Set<String>) -> Set<String> {
        Set(blocks.flatMap { pairKeys(in: $0) })
    }

    public static func canonicalizePairKey(_ key: String) -> String {
        if key.isEmpty || key.hasPrefix("/") {
            return key
        }
        return "/" + key
    }

    private func allGroupsDone(_ recipe: String, sections: [RecipeSection]) -> Bool {
        let keys = Self.pairKeys(for: recipe, in: sections)
        return !keys.isEmpty && keys.allSatisfy { blocks.contains($0) }
    }
}
